### ISSUE TITLE
gradle fetchDeps is platform-dependent

### ISSUE THREAD

**Issue by FliegendeWurst** [NixOS/nixpkgs#365375]
## Describe the bug

As observed in #353165, fetchDeps only resolves the dependencies needed for the build platform.
This leads to build failures later on if the same deps are used on other platforms.

## Steps To Reproduce

Use openjfx in a gradle project. This may apply to other dependencies too.

## Expected behavior

Just Works ™

## Screenshots

```
       >    > Could not find javafx-controls-17-linux-aarch64.jar (org.openjfx:javafx-controls:17).
       >      Searched in the following locations:
       >          https://repo.maven.apache.org/maven2/org/openjfx/javafx-controls/17/javafx-controls-17-linux-aarch64.jar
       >    > Could not find javafx-graphics-17-linux-aarch64.jar (org.openjfx:javafx-graphics:17).
       >      Searched in the following locations:
       >          https://repo.maven.apache.org/maven2/org/openjfx/javafx-graphics/17/javafx-graphics-17-linux-aarch64.jar
       >    > Could not find javafx-base-17-linux-aarch64.jar (org.openjfx:javafx-base:17).
       >      Searched in the following locations:
       >          https://repo.maven.apache.org/maven2/org/openjfx/javafx-base/17/javafx-base-17-linux-aarch64.jar
```

## Additional context

For this particular package it might be possible to use nixpkgs' openjfx package.

## Notify maintainers

gradle: @lorenzleutgeb @liff @NixOS/java 
mitm-cache: @chayleaf 

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by liff** [comment]
If I understand this correctly: given the way mitm-cache works, the `deps.json` cannot be populated with platform-dependent dependencies without actually running the build on all supported platforms?

So, in the case of keyboard-layout-editor, we’d need to collect `deps.json` on all platforms and somehow merge it?

**Comment by FliegendeWurst** [comment]
> So, in the case of keyboard-layout-editor, we’d need to collect deps.json on all platforms and somehow merge it?

Indeed, that would be the solution. Unless Gradle has some flag to specify the platform to build for.

**Comment by chayleaf** [comment]
Gradle builds scripts are Turing-complete, and dependency resolution is Turing-complete as well, the only way to remotely achieve anything like automatically fetching dependencies for all platforms would be static analysis of Gradle build scripts, which is definitely out of scope, the current integration took a lot of effort already. Please look at some of the existing Gradle packages - they run Gradle multiple times in the update script to fetch dependencies for all platforms.

**Comment by liff** [comment]
Taking an example from [jabref](https://github.com/NixOS/nixpkgs/blob/a33ea1ad5c6cb97a3208be54ab68599c4f36e79d/pkgs/applications/office/jabref/default.nix#L122-L127), this should work as for keyboard-layout-editor to resolve dependencies for all architectures:

```nix
  gradleUpdateScript = ''
    runHook preBuild

    gradle nixDownloadDeps -Dos.family=linux -Dos.arch=amd64
    gradle nixDownloadDeps -Dos.family=linux -Dos.arch=aarch64
    gradle nixDownloadDeps -Dos.name='mac os x' -Dos.arch=amd64
    gradle nixDownloadDeps -Dos.name='mac os x' -Dos.arch=aarch64
  '';
```

**Comment by tomodachi94** [comment]
It looks like this is specific to OpenJFX. I cannot find any reference to these properties in the gradle manual: https://docs.gradle.org/current/userguide/build_environment.html#gradle_properties_reference

**Comment by chayleaf** [comment]
> I cannot find any reference to these properties in the gradle manual

Gradle itself is a Java application that *mostly* doesn't have any platform-dependent behavior. Whenever there is platform-dependent behavior, it's the build scripts checking env vars and etc to figure out what dependencies to pick, what code to build, etc.

`os.name` and friends are system properties, see [Java docs](https://docs.oracle.com/javase/tutorial/essential/environment/sysprop.html). However, it doesn't mean this is how os detection is always done in Gradle. See, for example, [SIGNALD_TARGET](https://gitlab.com/signald/signald/-/blob/1583b4dfeaeec08bc34653ee55bf66731a2c6f75/build.gradle#L23) env var.

In case of OpenJFX, the official way is to use the OpenJFX plugin (see [this](https://github.com/trustin/os-maven-plugin/blob/master/src/main/java/kr/motd/maven/os/Detector.java)), but it's often done manually (don't have the time to dig for examples so [here's one from OpenJFX issues](https://github.com/openjfx/javafx-gradle-plugin/issues/94#issuecomment-754879310)).

Gradle itself uses `uname` behind the scenes (and signald above uses the Gradle implementation unless you override the env var), falling back to system properties.

In essence, `os.name`-like properties is the closest we have to overriding the OS, but it's not necessary unless the Gradle program uses native libs, and it's not sufficient unless you've looked at the build script and confirmed it.
